### PR TITLE
Add container count into resource quota

### LINKED ISSUE #6529: Setting an absolute cVM limit per VCH
VIC Engine - Any

User statement: As a customer of VIC, I would like set an absolute limit of cVMs per VCH, regardless of the resource pool limit set. This could be a parameter to vic-machine-{os} create.

Details: Resource limits are only applicable while cVMs are running. However a failed entrypoint and a poorly written plugin / script could keep retrying the create / run operation that would result in large number of stale (powered off) cVMs. 

Thanks.

Edit by @anchal-agrawal:

An additional acceptance criteria for this issue is to write test plans for CI and implement them. The tests can borrow logic from the ROBO tests for the container VM limit feature.

### PR DESCRIPTION
**Pull request by wjun**
Fixes #6529 
[specific ci=Group26-Quota.26-03-ContainerCount.robot]

### CODE REVIEW

File: cmd/vic-machine/configure/configure.go [hunk 1]
```diff
@@ -514,6 +515,15 @@ func (c *Configure) Run(clic *cli.Context) (err error) {
 		}
 	}
 
+	if c.ContainerCount != nil {
```
Comment (yuyangbj) [inline cmd/vic-machine/configure/configure.go]: The default value should be 0, not nil?

File: cmd/vic-machine/create/create.go [hunk 2]
```diff
@@ -702,6 +703,10 @@ func (c *Create) Run(clic *cli.Context) (err error) {
 		vchConfig.StorageQuota = quotaBytes
 	}
 
+	if c.ContainerCount != nil && *c.ContainerCount > 0 {
```
Comment (yuyangbj) [inline cmd/vic-machine/create/create.go]: The type of ContainerCount is int, so it cannot be nil forever?

File: cmd/vic-machine/configure/configure.go (hunk 1 above)
Comment (yuyangbj) [inline cmd/vic-machine/configure/configure.go]: Do we have a null pointer scenario? If so, should we set the ContainerCount to 0 for this case?

Comment (wjun) [reply cmd/vic-machine/configure/configure.go]: When we configure VCH, but do not touch container count, or any other options, these options will be nil. So not null indicates users have reset this option from command line.

File: cmd/vic-machine/create/create.go (hunk 2 above)
Comment (YanzhaoLi) [inline cmd/vic-machine/create/create.go]: Why this code differs than that in `configure` part ? Needn't set `vchConfig.ContainerCount = 0`?

File: lib/apiservers/engine/backends/container.go [hunk 3]
```diff
@@ -2134,8 +2138,34 @@ func (c *ContainerBackend) validateContainerLogsConfig(vc *viccontainer.VicConta
...
+		case *containers.GetContainerListInternalServerError:
+			return fmt.Errorf("Error invoking GetContainerList: %s", err.Payload.Message)
+
+		default:
+			return fmt.Errorf("Error invoking GetContainerList: %s", err.Error())
+		}
+	}
+
+	count := vchConfig.Cfg.ContainerCount
+	if count == len(containme.Payload) {
```
Comment (YanzhaoLi) [inline lib/apiservers/engine/backends/container.go]: Any chances that `count > len(containme.Payload)` ?

File: lib/apiservers/engine/backends/container.go [hunk 4]
```diff
@@ -667,6 +667,10 @@ func (c *ContainerBackend) ContainerCreate(config types.ContainerCreateConfig) (
 
 	var err error
 
+	if err = checkContainerCount(op, PortLayerClient()); err != nil {
```
Comment (YanzhaoLi) [inline lib/apiservers/engine/backends/container.go]: Should we do similar concurrent control as `checkStorageQuota `?

Comment (wjun) [reply lib/apiservers/engine/backends/container.go]: Good catch. Done.

File: lib/apiservers/engine/backends/container.go (hunk 3 above)
Comment (wjun) [reply lib/apiservers/engine/backends/container.go]: Good point. In parallel creations, there might be that case. Fixed.

File: cmd/vic-machine/create/create.go (hunk 2 above)
Comment (wjun) [reply cmd/vic-machine/create/create.go]: The default value during creation is 0.

File: lib/apiservers/engine/backends/container.go [hunk 5]
```diff
@@ -667,6 +667,11 @@ func (c *ContainerBackend) ContainerCreate(config types.ContainerCreateConfig) (
...
+		return containertypes.ContainerCreateCreatedBody{}, err
+	}
+	defer cache.ContainerCache().DecreaseContainerCount()
```
Comment (YanzhaoLi) [inline lib/apiservers/engine/backends/container.go]: What's the logic of this code?

Comment (wjun) [reply lib/apiservers/engine/backends/container.go]: checkContainerCount will reserve one container request by increase reservation number. When container creation finishes, we need to decrease the reservation number, as the container creation has finished.

**Pull request merged**